hnn_core/gui/gui.py

```diff
@@ -39,232 +59,401 @@
+    if rowcols_str != "":
+        rowcols = [int(_) for _ in rowcols_str.split(",") if _ != '']
+    else:
+        rowcols = []
 
-        if not init:
-            _update_plot_window(simulation_data, analysis_config, plot_key)
+    if init_vis_plot_tree:
+        vis_plot_tree['plots'] = {}
 
-    return grid
+    vis_plot_tree['kwargs'] = analysis_config['subplot_kwargs'].value
 
+    viz_window.clear_output()
+    with viz_window:
+        try:
+            kwargs = eval(f'dict({vis_plot_tree["kwargs"]})')
+        except Exception:
+            kwargs = {}
+        try:
+            fig, axes = plt.subplots(*rowcols,
+                                     figsize=(int(viz_layout.width[:-2]) / dpi,
+                                              int(viz_layout.height[:-2]) /
+                                              dpi),
+                                     dpi=dpi,
+                                     **kwargs)
+        except Exception:
+            fig, axes = plt.subplots(
+                *rowcols,
+                figsize=(int(viz_layout.width[:-2]) / dpi,
+                         int(viz_layout.height[:-2]) / dpi),
+                dpi=dpi,
+            )
+            # reset illegal parameters
+            analysis_config['subplot_kwargs'].value = ""
 
-def _change_plot_type(grid, layout, plot_idx, plot_type):
-    assert layout in _plot_options.values(), ValueError("Illegal layout type")
-    if layout == "L-R":
-        assert 0 <= plot_idx < 2
-        grid[layout].children[plot_idx].children[0].value = plot_type
-        pass
-    elif layout == "U-D":
-        assert 0 <= plot_idx < 2
-        grid[layout].children[plot_idx].children[0].value = plot_type
+
+        if issubclass(type(axes), matplotlib.axes.SubplotBase):
+            idx_axes = [(0, axes)]
+        else:
+            idx_axes = [(idx, ax) for idx, ax in enumerate(axes.flat)]
+        for idx, ax in idx_axes:
+            if init_vis_plot_tree:
+                ax.set_title(f"Ax {idx}")
+                vis_plot_tree['plots'][idx] = []
+            else:
+                viz_configs = vis_plot_tree['plots'][idx]
+                for idx_overlay, viz_config in enumerate(viz_configs):
+                    sim_name = viz_config['sim_name']
+                    sinle_sim = simulation_data[sim_name]
+                    plot_type = viz_config['plot_type']
+                    plot_config = viz_config['config']
+                    _update_ax(ax, sinle_sim, plot_type, plot_config)
+
+
+        plt.tight_layout()
+        plt.show(fig)
+
+
+def _gen_plot_specific_analysis(plot_data):
+    """Generate config for a specific subplots"""
+    analysis_style = {'description_width': '200px'}
+    w_max_spectral_frequency = FloatText(
+        value=plot_data['config']['max_spectral_frequency'],
+        description='Max Spectral Frequency (Hz):',
+        disabled=False,
+        style=analysis_style)
+
+    def _handle_sf_change(b):
```

Comment:
what does `sf` stand for?

Reply:
spectral_frequency, i use a short name there just because it's a very simple function and the definition is very close to the caller. Another day wondering why there's no multiline anonymous function in Python...